### PR TITLE
WEB-6552: Remove dependence on single iin provider & add SLA

### PR DESCRIPTION
**Pull request by jamescogley**
#### What are the links to relevant tickets?

https://hxshortbreaks.atlassian.net/browse/WEB-6552
#### What does this PR do?

Rather than requesting iin lookups from each provider in turn we make requests to each provider and wait for the first one to successfully return the information we need.

If all the providers fail we fall back onto a RegEx lookup that will at least should be able to identify the brand of the card.
#### What functional/unit tests does this PR have?

No new tests added.
#### Screenshots (if appropriate)

---
#### Quality checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit test coverage.
- [ ] I've updated documentation with any changes to procedures.
- [x] I've checked this work against the requirements of the Jira.
- [x] This change passes all necessary tests (cross browser, automated or otherwise).
- [x] I am ready for this to be code reviewed, merged and tested.
#### Reviewer 1 @philmeehan
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!
#### Reviewer 2 @kevinhodges
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

### CODE REVIEW

File: configs/providers.js [hunk 1]
```diff
@@ -23,10 +23,10 @@ module.exports = [
 				iin: returnedData.bin,
 				brand: returnedData.brand,
 				issuer: returnedData.bank,
-				type: returnedData.card_type,
+				type: ( returnedData.card_type ? returnedData.card_type : nullValue ),
```
Comment (philmeehan) [inline configs/providers.js]:
```
type: returnedData.card_type || nullValue,
```
is a bit more succinct

File: lib/iinChecker.js [hunk 2]
```diff
@@ -132,18 +73,178 @@
...
+          // now clean up any requests so we're not waiting around
+          clearRequestQueue();
+          // everything else has failed and we're left with the RegEx result - let's just send that back
+          listenerCallback( null, cardInfo );
+          return true;
+        }
+        return true;
+      });
+
+      emitter.on('returnError', function( error, iinToLookup ) {
```
Comment (philmeehan) [inline lib/iinChecker.js]: Is this the right behaviour for an error?

File: test/cachingTest.js [hunk 3]
```diff
@@ -70,12 +70,12 @@ describe( 'iinChecker()', function() {
             throw ( err );
           } else {
             expect( result ).to.deep.equal( cardDetails );
-            expect( setCacheStub.calledOnce ).to.be.true;
+            expect( setCacheStub.calledOnce ).to.equal(true);
```
Comment (philmeehan) [inline test/cachingTest.js]: If you use dirty-chai you can keep these as 
```
expect( setCacheStub.calledOnce ).to.be.true();
```

We have used this quite a bit in other repos

File: lib/iinChecker.js [hunk 4]
```diff
@@ -46,13 +50,7 @@
...
 
     this.isCacheable = function( cardInfo ) {
-      if ( this.options.cache &&
-          cardInfo.brand && cardInfo.brand !== this.brands.UNKNOWN &&
-          cardInfo.type && cardInfo.type !== this.nullValue
-      ) {
-        return true;
-      }
-      return false;
+      return (this.options.cache && cardInfo.brand && cardInfo.brand !== this.brands.UNKNOWN && cardInfo.type && cardInfo.type !== this.nullValue);
```
Comment (kevinhodges) [inline lib/iinChecker.js]: I like this better than what was before but it's not that readable. One suggestion I have is to make it completely verbose and do something like this, totally up to you though
```
function checkCard( cardInfo ) {
	var self = this;
	// no cache so not cacheable
	if( !self.options.cache ) {
		return false;
	}
	// don't cache cards we don't know the brand for
	if( !cardInfo.brand || cardInfo.brand === self.brands.UNKNOWN ) {
		return false;
	}

	// don't cache cards we don't know the type of
	if( !cardInfo.type || cardInfo.type === this.nullValue ) {
		return false;
	}

	// cacheable!
	return true;

}
```

File: lib/iinChecker.js [hunk 5]
```diff
@@ -3,15 +3,19 @@
...
   var iinChecker = function( iinOptions ) {
+    // Load all of our providers into an array that we can loop over.
+    this.providers = require( '../configs/providers' );
+
     // default options
     this.options = {
       language: 'en',
-      cache: false
+      cache: false,
+      timeout: 1000
```
Comment (kevinhodges) [inline lib/iinChecker.js]: do we know how long our requests take on average? Can we tell from Graphite?

**Comment by philmeehan** [comment]
I get failing tests locally as q isn't in the devDependencies

File: lib/iinChecker.js [hunk 6]
```diff
@@ -132,18 +73,178 @@
...
+      // listen to the "regexResult" event - we have a result from
+      emitter.on('regexResult', function( cardInfo ) {
+        if ( cardInfo.iin === iin ) {
+          // we don't want to worry about any more events
+          stopListening();
+          // now clean up any requests so we're not waiting around
+          clearRequestQueue();
+          // everything else has failed and we're left with the RegEx result - let's just send that back
+          listenerCallback( null, cardInfo );
+          return true;
```
Comment (kevinhodges) [inline lib/iinChecker.js]: redundant `return true'` here

**Comment by philmeehan** [comment]
Also there are a lot of untested branches in the code within the emitted events. Would it be possible to add a few tests to catch a few more of these branches? 

The istanbul coverage report highlights them as E

File: lib/iinChecker.js (hunk 5 above)
Comment (philmeehan) [inline lib/iinChecker.js]: Think Reuben is adding metrics as part of his work

File: lib/iinChecker.js [hunk 7]
```diff
@@ -132,18 +73,178 @@
...
+      }
+    }
+
+    // Bind our listeners here so we can handle events
+    function bindListeners( listenerCallback ) {
+      // we need to keep track of any requests we have
+      var currentRequests = [];
+
+      function stopListening() {
+        emitter.removeAllListeners('requestStarted')
```
Comment (kevinhodges) [inline lib/iinChecker.js]: as you have a reference to `emitter` you can simply call `emitter.removeAllListeners();`

**Comment by jamescogley** [comment]
@philmeehan I'll add Q in as a dev-dependancy and see about getting some tests to cover the untested branches

File: lib/iinChecker.js [hunk 8]
```diff
@@ -132,18 +73,178 @@
...
+          .removeAllListeners('requestComplete')
+          .removeAllListeners('requestFailed')
+          .removeAllListeners('regexResult')
+          .removeAllListeners('returnError');
+      }
+
+      function clearRequestQueue() {
```
Comment (kevinhodges) [inline lib/iinChecker.js]: i've never used it before but would `_.invokeMap(currentRequests, 'abort');` work here?

File: configs/providers.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 module.exports = [
 	{
 		name: "RIBBON",
-		domain: "https://bins.ribbon.co",
+		domain: "http://bins.ribbon.co",
```
Comment (hpoom) [inline configs/providers.js]: Has ribbon really moved to http? If SSL is still an option we should favour https over http.

**Comment by hpoom** [comment]
@jamescogley It might be worth checking issue #32 is this resolved by the PR?

If not, not worried, but just checking if that needs to be address separately.

File: lib/iinChecker.js (hunk 6 above)
Comment (notABluesSinger) [inline lib/iinChecker.js]: should perhaps `return listenerCallback( null, cardInfo );`

**Comment by notABluesSinger** [comment]
**Pedant Alert:**
Appears to be a mixture of spacing between parentheses. Can we standardise on one style throughout please.

**Comment by notABluesSinger** [comment]
I think we need to either add a `.travis.yml` and get the tests passing and remove the Circle integration, or remove Travis.
As OSS, my vote is for Travis.

File: configs/providers.js (hunk 9 above)
Comment (jamescogley) [inline configs/providers.js]: Ribbon appears to be Payout now with bins.ribbon.co redirecting to bins.payout.com.  The ssl certificate for bins.ribbon.co is no longer valid - there also appears to be a 1000 request daily limit now as well.

File: lib/iinChecker.js (hunk 8 above)
Comment (jamescogley) [inline lib/iinChecker.js]: looks like it should, will give it a go

File: lib/iinChecker.js (hunk 7 above)
Comment (jamescogley) [inline lib/iinChecker.js]: good point!

File: lib/iinChecker.js (hunk 4 above)
Comment (jamescogley) [inline lib/iinChecker.js]: Looking at it, it is a bit too much for one line.

**Comment by jamescogley** [comment]
@notABluesSinger do we have any examples of a `.travis.yml ` I can work with?

**Comment by notABluesSinger** [comment]
https://github.com/holidayextras/culture/blob/master/ci-best-practices.md

File: configs/providers.js (hunk 9 above)
Comment (notABluesSinger) [inline configs/providers.js]: Do you think that moving away from `https` is the correct solution here?

Comment (jamescogley) [inline configs/providers.js]: I will updated the Ribbon provider with the new API details which supports https

**Comment by jamescogley** [comment]
@hpoom unfortunately this PR won't help with #32

File: lib/iinChecker.js (hunk 5 above)
Comment (kevinhodges) [inline lib/iinChecker.js]: 👍

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+node_js:
+  - 0.10
+cache:
+  directories:
+  - node_modules
+before_script:
+  - npm install -g grunt-cli
+  - npm install -g mocha
+after_script:
+  - istanbul cover _mocha --report lcovonly && cat ./coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js
```
Comment (hxpaul) [inline .travis.yml]: do you know why we use istanbul and coveralls? istanbul can do it all I think

Comment (hpoom) [inline .travis.yml]: We added coveralls to give us and other work on this visibility on coverage beyond the terminal. Also be able to link to it in PR's like the comment above from coveralls.

I might be wrong but I thought istanbul was command line only and has no links or hooks into GitHub or a web dashboard?

Comment (hxpaul) [inline .travis.yml]: ok I see this is all approved so ignore me, but i'd like to see consistency across our repos... maybe that means adding coveralls in everywhere else

Comment (hpoom) [inline .travis.yml]: Coveralls also give a nice view of historical coverage. So you can track coverage over time. We did have it in quite a few repos, but I think now it is only on Open Source.

Comment (hxpaul) [inline .travis.yml]: I haven't noticed it on any other repos yet, maybe I've been focussed on the closed repos. We are now tracking our coverage on graphite with for example https://github.com/holidayextras/revolver/blob/staging/circle.yml#L23

**Pull request merged**